src/main/java/websocket/WSManager.java

```diff
@@ -0,0 +1,160 @@
+    // HashMap for keeping track of sent requests (Tag -> Request)
+    HashMap<Long, Request> requestHashMap;
+    // HashMap for registered notification handlers (Resource.Method -> Handler)
+    HashMap<String, INotificationHandler> notificationHandlerHashMap;
+    // WebSocket connection
+    WSConnection socket;
+    // Jackson Mapper
+    private ObjectMapper mapper = new ObjectMapper();
+
+    public WSManager(ConnectionConfig config) {
+        this.notificationHandlerHashMap = new HashMap<>();
+        this.requestHashMap = new HashMap<>();
+        this.socket = new WSConnection(config);
+    }
+
+    // used for testing
+    WSManager(WSConnection socket) {
+        this.notificationHandlerHashMap = new HashMap<>();
+        this.requestHashMap = new HashMap<>();
+        this.socket = socket;
+    }
+
+    /**
+     * Register an INotificationHandler to receive notifications from the server.
+     *
+     * @param type
+     * @param handler
+     */
+    public void registerNotificationHandler(String type, INotificationHandler handler) {
+        notificationHandlerHashMap.put(type, handler);
+    }
+
+    /**
+     * Deregister the INotificationHandler saved for the given type of notification.
+     *
+     * @param type
+     */
+    public void deregisterNotificationHandler(String type) {
+        notificationHandlerHashMap.remove(type);
+    }
+
+    /**
+     * Send a request over the WebSocket connection.
+     *
+     * @param request
+     */
+    public void sendRequest(Request request) throws ConnectException {
+        if (socket.getState() == WSConnection.State.CREATED || socket.getState() == WSConnection.State.ERROR) {
+            try {
+                socket.connect();
+            } catch (Exception e) {
+                logger.error("WebSocket connection could not connect.");
+                throw new ConnectException("Could not connect to WebSocket.");
+            }
+        }
+        String messageText;
+        try {
+            messageText = mapper.writeValueAsString(request);
+        } catch (JsonProcessingException e) {
+            logger.error("Could not map request to Json string: " + request);
+            return;
+        }
+        socket.enqueueMessage(messageText);
+    }
+
+    @Override
+    public void handleMessage(String message) {
+        ServerMessageWrapper wobject;
+        try {
+            wobject = mapper.readValue(message, ServerMessageWrapper.class);
+        } catch (IOException e) {
+            logger.error("Malformed message from server: " + message);
+            return;
+        }
+        switch (wobject.getType()) {
+            case ServerMessageWrapper.TYPE_NOTIFICATION:
+                handleNotification(wobject);
+                break;
+            case ServerMessageWrapper.TYPE_RESPONSE:
+                handleResponse(wobject);
+                break;
+        }
+    }
+
+    private void handleNotification(ServerMessageWrapper wobject) {
```

Comment:
the fact that this is named the same as the message in IMessageHandler is confusing